src/main/java/teammates/storage/entity/Account.java

```diff
@@ -124,12 +121,49 @@ public void setCreatedAt(Date createdAt) {
         this.createdAt = createdAt;
     }
 
+    /**
+     * Fetches the student profile from the datastore the first time this is called. Returns null if student profile was
+     * explicitly set to null (e.g. when the student profile is intentionally not retrieved). If a shell student profile
+     * with an empty Google ID was set, simply returns this shell student profile without interacting with the datastore.
+     */
     public StudentProfile getStudentProfile() {
-        return this.studentProfile;
+        if (!isStudentProfileEnabled) {
+            return null;
+        }
+        if (localStudentProfile != null && localStudentProfile.getGoogleId().isEmpty()) { // only in local attribute tests
+            return localStudentProfile;
+        }
+        if (studentProfile == null) {
+            return null;
+        }
+        return studentProfile.get();
     }
 
+    /**
+     * Sets a reference to {@code studentProfile} which subsequent calls to {@code getStudentProfile()} will use to fetch
+     * from. To disable this behaviour (e.g. when the student profile is intentionally not retrieved), set to null. If a
+     * shell student profile with an empty Google ID is set, subsequent calls to {@code getStudentProfile()} will simply
+     * return the shell student profile without interacting with the datastore.
+     */
     public void setStudentProfile(StudentProfile studentProfile) {
-        this.studentProfile = studentProfile;
+        if (studentProfile == null) {
+            setIsStudentProfileEnabled(false);
+            return;
+        }
+        setIsStudentProfileEnabled(true);
+        if (studentProfile.getGoogleId().isEmpty()) { // only in local attribute tests
```

Comment:
In principle, this PR should avoid modifying tests as much as possible. Instead, it should show that even with the code changes in the Storage layer, all existing tests still pass.